src/canisters/frontend/ic-certified-assets/src/state_machine.rs

```diff
@@ -830,117 +1018,119 @@ impl From<StableState> for State {
+        asset_hashes.delete(key_path.asset_hash_path_root_v2().as_vec());
+        asset_hashes.delete(key_path.asset_hash_path_v1().as_vec());
+        if key == INDEX_FILE {
+            asset_hashes.delete(&[
+                NestedTreeKey::String("http_expr".into()),
+                NestedTreeKey::String("<*>".into()),
+            ]);
         }
-        enc.certified = true;
     }
 }
 
-fn witness_to_header(witness: HashTree, certificate: &[u8]) -> HeaderField {
-    use ic_certified_map::labeled;
-
-    let hash_tree = labeled(b"http_assets", witness);
-    let mut serializer = serde_cbor::ser::Serializer::new(vec![]);
-    serializer.self_describe().unwrap();
-    hash_tree.serialize(&mut serializer).unwrap();
-
-    (
-        "IC-Certificate".to_string(),
-        String::from("certificate=:")
-            + &base64::encode(certificate)
-            + ":, tree=:"
-            + &base64::encode(&serializer.into_inner())
-            + ":",
-    )
-}
-
-fn merge_hash_trees<'a>(lhs: HashTree<'a>, rhs: HashTree<'a>) -> HashTree<'a> {
-    use HashTree::{Empty, Fork, Labeled, Leaf, Pruned};
-
-    match (lhs, rhs) {
-        (Pruned(l), Pruned(r)) => {
-            if l != r {
-                panic!("merge_hash_trees: inconsistent hashes");
-            }
-            Pruned(l)
-        }
-        (Pruned(_), r) => r,
-        (l, Pruned(_)) => l,
-        (Fork(l), Fork(r)) => Fork(Box::new((
-            merge_hash_trees(l.0, r.0),
-            merge_hash_trees(l.1, r.1),
-        ))),
-        (Labeled(l_label, l), Labeled(r_label, r)) => {
-            if l_label != r_label {
-                panic!("merge_hash_trees: inconsistent hash tree labels");
-            }
-            Labeled(l_label, Box::new(merge_hash_trees(*l, *r)))
+fn insert_new_response_hashes_for_encoding(
+    asset_hashes: &mut AssetHashes,
+    enc: &AssetEncoding,
+    affected_keys: &Vec<String>,
+) {
+    for key in affected_keys {
+        let key_path = AssetPath::from(&key);
+        let v1_path = key_path.asset_hash_path_v1();
+        if asset_hashes.get(v1_path.as_vec()).is_none() {
+            // v1 can only certify one encoding, therefore we only insert the first encoding as defined by encoding_certification_order()
+            asset_hashes.insert(v1_path.as_vec(), enc.sha256.into());
         }
-        (Empty, Empty) => Empty,
-        (Leaf(l), Leaf(r)) => {
-            if l != r {
-                panic!("merge_hash_trees: inconsistent leaves");
+        for status_code in STATUS_CODES_TO_CERTIFY {
+            if let Some(hash_path) = enc.asset_hash_path_v2(&key_path, status_code) {
+                asset_hashes.insert(hash_path.as_vec(), Vec::new());
+            } else {
+                unreachable!(
+                    "Could not create a hash path for a status code {} and key {} - did you forget to add a response hash for this status code?",
+                    status_code, &key
+                );
```

Comment:
I guess this really is supposed to be unreachable.  A small suggestion to tie this to the likely culprit (compute_response_hashes).  When I read the verb "add" I thought maybe it had to be defined in the asset configuration somewhere.
```suggestion
                unreachable!(
                    "Could not create a hash path for a status code {} and key {} - did you forget to compute a response hash for this status code?",
                    status_code, &key
                );
```